ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -104,6 +104,20 @@ public RaftPeerRole getCurrentRole() {
       return getRole().getCurrentRole();
     }
 
+    @Override
+    public boolean isLeader() {
```

Comment:
Assume that the leader won't step down if it has lost the leader lease.

If the assumption is correct, then it is better to change isLeaderRead() instead of isLeader().  The client will see LeaderNotReadyException and retry with the same leader.

If we change isLeader(),  the client will see NotLeaderException and retry with a different peer.

Reply:
> Assume that the leader won't step down if it has lost the leader lease.

@szetszwo Yes, leader won't step down if it has lost the leader lease.
Because I'm afraid if step down, leader election will happen more frequent. If not step down, leader lost the lease maybe can get the lease again in future. If leader actully need to step down, we do achieve it by checkLeaderShip, so there is a buffer time between lost lease and step down for leader to get lease again.  Step down or not, I think it's still under discussion.

> If we change isLeader(), the client will see NotLeaderException and retry with a different peer.

Yes, but user maybe prefer to call isLeader, maybe retry with a different peer does not matter ? not sure.

Reply:
> If leader actully need to step down, we do achieve it by checkLeaderShip, ...

Yes.  I agree.

> ... user maybe prefer to call isLeader, ...

Not the user code.  It is the code in RaftServerImpl.checkLeaderState.
```
//RaftServerImpl.checkLeaderState
    if (!getInfo().isLeader()) {
      NotLeaderException exception = generateNotLeaderException();
      final RaftClientReply reply = newExceptionReply(request, exception);
      return RetryCacheImpl.failWithReply(reply, entry);
    }
    if (!getInfo().isLeaderReady()) {
      final CacheEntry cacheEntry = retryCache.getIfPresent(ClientInvocationId.valueOf(request));
      if (cacheEntry != null && cacheEntry.isCompletedNormally()) {
        return cacheEntry.getReplyFuture();
      }
      final LeaderNotReadyException lnre = new LeaderNotReadyException(getMemberId());
      final RaftClientReply reply = newExceptionReply(request, lnre);
      return RetryCacheImpl.failWithReply(reply, entry);
    }
```

Reply:
> Not the user code. It is the code in RaftServerImpl.checkLeaderState.

@szetszwo Yes.  But I'm not sure whether it's interface friendly, maybe user is likely to check isLeader, not isLeaderReady. Just a discussion, I'm not sure.

Reply:
Sorry that I was not clear.  By "user", I actually mean RaftClient.  RaftServerImpl.checkLeaderState is an internal private method.  It is called by all the leader methods such as submitClientRequestAsync(..). RaftClient will get a NotLeaderException when isLeader() returns false, or LeaderNotReadyException when isLeaderReady() returns false.

If the leader is not going to step down, it is better to change isLeaderReady().  The leader may regain the leader lease so that RaftClient should retry with the same leader.

Reply:
@szetszwo Oh, sorry. By "user", I mean OM or SCM, because they need to check isLeader to decide whether do some thing, for example only leader SCM send command to datanode, so SCM need to know whether it's leader. Then the problem becomes SCM call which method: isLeader() or isLeaderReady(). From the name of method, maybe SCM is likely to use isLeader().

Reply:
@szetszwo After thinking twice. I agree with you. We should change isLeaderReady().  Thanks the explanation.

Reply:
Sure, thanks.

Reply:
@runzhiwang  Thanks for the tag. I have following suggestions:

1. I agree that leader should sort the `sendTsWithAck` increasingly, and iterate from large to small to pick the `sendTsWithAck` that majority of the followers(including himself) has replied the leader as the `startOfLease` .
2. We can use `LeaderStateImpl#checkLeadership` to calculate and maintain the lease, and step down when lease is expired. `isLeader()` might be called hundreds or thousands times per second, which would be a potential waste to resource. `LeaderStateImpl#checkLeadership` can make leader step down if lease is expired, which is more suitable for the semantics of lease.
3. We may not need the `LEADER_LEASE_TIMEOUT_RATIO_KEY ` at the very beginning. What we worry about is the lease is too short to make leader unstable, and we can revisit the `clock drift bound` feature later.
4. To maintain the lease, the follower should withhold its vote if its has replied the AppendEntriesRequest, even it meets a RequestVoteRequest with higher term and newer raft log. Seems the RequestVote logic is not touched in this PR ?
5. I suggest that leader should step down when the maintained lease is expired. From the view of App layer, `isLeaderReady()` means raft server has become leader, yet the `appliedLogIndex` of StateMachine has not reach the `lastLogIndex`, As soon as `isLeaderReady()` return true, the App layer would expect leader not return false during the whole term.

Reply:
> To maintain the lease, the follower should withhold its vote if its has replied the AppendEntriesRequest, even it meets a RequestVoteRequest with higher term and newer raft log. Seems the RequestVote logic is not touched in this PR ?

@GlenGeng This will be done in Pre-vote PR: https://github.com/apache/incubator-ratis/pull/161